plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMigrateProcedure.java

```diff
@@ -76,6 +77,43 @@ public void testMigrateTable(IcebergFileFormat fileFormat)
         assertUpdate("DROP TABLE " + tableName);
     }
 
+    @Test(dataProvider = "fileFormats")
+    public void testMigrateTableWithComplexType(IcebergFileFormat fileFormat)
+    {
+        String tableName = "test_migrate_complex_" + randomNameSuffix();
+        String hiveTableName = "hive.tpch." + tableName;
+        String icebergTableName = "iceberg.tpch." + tableName;
+
+        assertUpdate("CREATE TABLE " + hiveTableName + " WITH (format='" + fileFormat + "') AS SELECT 1 x, array[2, 3] a, CAST(map(array['key1'], array['value1']) AS map(varchar, varchar)) b, CAST(row(1) AS row(d integer)) c", 1);
+        assertQueryFails("SELECT * FROM " + icebergTableName, "Not an Iceberg table: .*");
+
+        assertUpdate("CALL iceberg.system.migrate('tpch', '" + tableName + "')");
+
+        assertThat((String) computeScalar("SHOW CREATE TABLE " + icebergTableName))
+                .contains("format = '%s'".formatted(fileFormat));
+
+        @Language("SQL") String firstRow = "VALUES (" +
+                "1, " +
+                "ARRAY[2, 3], " +
+                "CAST(map(ARRAY['key1'], ARRAY['value1']) AS map(varchar, varchar)), " +
+                "CAST(row(1) AS row(d integer)))";
+        assertThat(query("SELECT * FROM " + icebergTableName))
+                .matches(firstRow);
+        assertQuery("SELECT count(*) FROM " + icebergTableName, "VALUES 1");
+
+        @Language("SQL") String secondRow = " VALUES (" +
+                "2, " +
+                "ARRAY[4, 5], " +
+                "CAST(map(ARRAY['key2'], ARRAY['value2']) AS map(varchar, varchar)), " +
+                "CAST(row(2) AS row(d integer)))";
+        assertUpdate("INSERT INTO " + icebergTableName + secondRow, 1);
+        assertQuery("SELECT count(*) FROM " + icebergTableName, "VALUES 2");
+        assertThat(query("SELECT * FROM " + icebergTableName))
+                .matches(firstRow + " UNION ALL " + secondRow);
+
+        assertUpdate("DROP TABLE " + tableName);
+    }
+
     @DataProvider
```

Comment:
Consider adding the following test cases to your PR

```
    @Test
    public void testMigrateTableParquetSchemaEvolution()
            throws Exception
    {
        String randomNameSuffix = randomNameSuffix();
        String tableNameOneColumn = "test_migrate_one_column_" + randomNameSuffix;
        String tableNameTwoColumns = "test_migrate_two_columns_" + randomNameSuffix;
        String hiveTableNameOneColumn = "hive.tpch." + tableNameOneColumn;
        String hiveTableNameTwoColumns = "hive.tpch." + tableNameTwoColumns;
        String icebergTableNameTwoColumns = "iceberg.tpch." + tableNameTwoColumns;

        assertUpdate("CREATE TABLE " + hiveTableNameOneColumn + " WITH (format = 'PARQUET') AS SELECT 1 col1", 1);
        assertUpdate("CREATE TABLE " + hiveTableNameTwoColumns + " WITH (format = 'PARQUET') AS SELECT 2 col1, CAST(row(10, 20) AS row(x integer, y integer)) AS nested", 1);

        // Copy the parquet file containing only one column to the table with two columns
        Path tableNameOneColumnLocation = Path.of("%s/tpch/%s".formatted(dataDirectory, tableNameOneColumn));
        Path tableNameTwoColumnsLocation = Path.of("%s/tpch/%s".formatted(dataDirectory, tableNameTwoColumns));
        try (Stream<Path> files = Files.list(tableNameOneColumnLocation)) {
            Path file = files.filter(path -> !path.getFileName().toString().startsWith(".")).collect(onlyElement());
            Files.copy(file, tableNameTwoColumnsLocation.resolve(file.getFileName()));
        }

        assertUpdate("CALL iceberg.system.migrate('tpch', '" + tableNameTwoColumns + "')");

        assertThat(query("SELECT * FROM " + icebergTableNameTwoColumns))
                .skippingTypesCheck()
                .matches("VALUES (1, CAST(null AS row(x integer, y integer))), (2, row(10, 20))");

        assertUpdate("INSERT INTO " + icebergTableNameTwoColumns + " VALUES (3, row(100, 200))", 1);
        assertThat(query("SELECT * FROM " + icebergTableNameTwoColumns))
                .skippingTypesCheck()
                .matches("VALUES (1, CAST(null AS row(x integer, y integer))), (2, row(10, 20)), (3, row(100, 200))");

        assertUpdate("DROP TABLE " + icebergTableNameTwoColumns);
    }

    @Test
    public void testMigrateTableParquetRowColumnSchemaEvolution()
            throws Exception
    {
        String randomNameSuffix = randomNameSuffix();
        String tableNameRowOneColumn = "test_migrate_row_one_column_" + randomNameSuffix;
        String tableNameRowTwoColumns = "test_migrate_row_two_columns_" + randomNameSuffix;
        String hiveTableNameRowOneColumn = "hive.tpch." + tableNameRowOneColumn;
        String hiveTableNameRowTwoColumns = "hive.tpch." + tableNameRowTwoColumns;
        String icebergTableNameRowTwoColumns = "iceberg.tpch." + tableNameRowTwoColumns;

        assertUpdate("CREATE TABLE " + hiveTableNameRowOneColumn + " WITH (format = 'PARQUET') AS SELECT CAST(row(1) AS row(x integer)) as nested", 1);
        assertUpdate("CREATE TABLE " + hiveTableNameRowTwoColumns + " WITH (format = 'PARQUET') AS SELECT CAST(row(10, 20) AS row(x integer, y integer)) AS nested", 1);

        Path tableNameRowOneColumnLocation = Path.of("%s/tpch/%s".formatted(dataDirectory, tableNameRowOneColumn));
        Path tableNameRowTwoColumnsLocation = Path.of("%s/tpch/%s".formatted(dataDirectory, tableNameRowTwoColumns));
        try (Stream<Path> files = Files.list(tableNameRowOneColumnLocation)) {
            Path file = files.filter(path -> !path.getFileName().toString().startsWith(".")).collect(onlyElement());
            Files.copy(file, tableNameRowTwoColumnsLocation.resolve(file.getFileName()));
        }

        assertUpdate("CALL iceberg.system.migrate('tpch', '" + tableNameRowTwoColumns + "')");

        assertThat(query("SELECT * FROM " + icebergTableNameRowTwoColumns))
                .skippingTypesCheck()
                .matches("VALUES row(CAST((1,null) AS row(x integer, y integer))), row(row(10, 20))");

        assertUpdate("INSERT INTO " + icebergTableNameRowTwoColumns + " VALUES (row(row(100, 200)))", 1);
        assertThat(query("SELECT * FROM " + icebergTableNameRowTwoColumns))
                .skippingTypesCheck()
                .matches("VALUES row(CAST((1, null) AS row(x integer, y integer))), row(row(10, 20)), row(row(100, 200))");

        assertUpdate("DROP TABLE " + icebergTableNameRowTwoColumns);
    }

```

Reply:
handsome